std/machines/hash/keccakf16.asm

```diff
@@ -0,0 +1,576 @@
+    });
+
+    // // A''[x, y] = xor(B[x, y], andn(B[x + 1, y], B[x + 2, y])).
+    // for y in 0..5 {
+    //     for x in 0..5 {
+    //         let get_bit = |z| {
+    //             let andn = andn_gen::<AB::Expr>(
+    //                 local.b((x + 1) % 5, y, z).into(),
+    //                 local.b((x + 2) % 5, y, z).into(),
+    //             );
+    //             xor_gen::<AB::Expr>(local.b(x, y, z).into(), andn)
+    //         };
+
+    //         for limb in 0..U64_LIMBS {
+    //             let computed_limb = (limb * BITS_PER_LIMB..(limb + 1) * BITS_PER_LIMB)
+    //                 .rev()
+    //                 .fold(AB::Expr::zero(), |acc, z| acc.double() + get_bit(z));
+    //             builder.assert_eq(computed_limb, local.a_prime_prime[y][x][limb]);
+    //         }
+    //     }
+    // }
+
+    let andn: expr, expr -> expr = |a, b| (1 - a) * b;
+
+    array::new(100, |i| {
+        let y = i / 20;
+        let x = (i / 4) % 5;
+        let limb = i % 4;
+
+        let get_bit: int -> expr = |z| {
+            xor(b(x, y, z), andn(b((x + 1) % 5, y, z), b((x + 2) % 5, y, z)))
+        };
+        let computed_limb: expr = utils::fold(16, |z| get_bit((limb + 1) * 16 - 1 - z), 0, |acc, e| (acc * 2 + e));
+        computed_limb - a_prime_prime[i] = 0
+    });
+
+    // pub fn b(&self, x: usize, y: usize, z: usize) -> T {
+    //     debug_assert!(x < 5);
+    //     debug_assert!(y < 5);
+    //     debug_assert!(z < 64);
+
+    //     // B is just a rotation of A', so these are aliases for A' registers.
+    //     // From the spec,
+    //     //     B[y, (2x + 3y) % 5] = ROT(A'[x, y], r[x, y])
+    //     // So,
+    //     //     B[x, y] = f((x + 3y) % 5, x)
+    //     // where f(a, b) = ROT(A'[a, b], r[a, b])
+    //     let a = (x + 3 * y) % 5;
+    //     let b = x;
+    //     let rot = R[a][b] as usize;
+    //     self.a_prime[b][a][(z + 64 - rot) % 64]
+    // }
+
+    let b: int, int, int -> expr = |x, y, z| {
+        let a: int = (x + 3 * y) % 5;
+        let rot: int = R[a * 5 + x]; // b = x
+        a_prime[x * 320 + a * 64 + (z + 64 - rot) % 64]
+    };
+
+    // // A'''[0, 0] = A''[0, 0] XOR RC
+    // for limb in 0..U64_LIMBS {
+    //     let computed_a_prime_prime_0_0_limb = (limb * BITS_PER_LIMB
+    //         ..(limb + 1) * BITS_PER_LIMB)
+    //         .rev()
+    //         .fold(AB::Expr::zero(), |acc, z| {
+    //             builder.assert_bool(local.a_prime_prime_0_0_bits[z]);
+    //             acc.double() + local.a_prime_prime_0_0_bits[z]
+    //         });
+    //     let a_prime_prime_0_0_limb = local.a_prime_prime[0][0][limb];
+    //     builder.assert_eq(computed_a_prime_prime_0_0_limb, a_prime_prime_0_0_limb);
+    // }
+
+    array::map(a_prime_prime_0_0_bits, |i| force_bool(i));
+
+    array::new(4, |limb| {
+        let computed_a_prime_prime_0_0_limb = utils::fold(16, |z| a_prime_prime_0_0_bits[(limb + 1) * 16 - 1 - z], 0, |acc, e| (acc * 2 + e));
```

Comment:
The fold with `acc * 2, e` could also be a helper function `bits_to_value` or something.

Reply:
Created `bits_to_limb_be`